text/0000-slice-ptr-range.md

```diff
@@ -0,0 +1,150 @@
+- Feature Name: `slice_ptr_range`
+- Start Date: 2019-10-25
+- RFC PR: None yet
+- Rust Issue: None yet
+
+# Summary
+[summary]: #summary
+
+Add `.as_ptr_range()` and `.as_mut_ptr_range()` to slices, which give
+both the start and the (one-past-the-)end pointer of a slice.
+(As a `Range<*const T>` or `Range<*mut T>`.)
+
+# Motivation
+[motivation]: #motivation
+
+Many C and C++ APIs use a range of pointers instead of pointer and size to
+refer to a slice of items. These functions makes it easier to obtain those two
+pointers.
+
+The alternative right now is to either use:
+
+```rust
+let start = slice.as_ptr();
+let end = slice[slice.len()..].as_ptr();
+some_ffi_function(start, end);
+```
+
+Or to use the (unsafe) `add` function to add the `len()` to `as_ptr()`:
+The current documentation for raw pointers shows a few examples of
+`vec.as_ptr().add(vec.len())`.
+
+Both are not ideal.
+
+With these new functions, it'd be:
+
+```rust
+let r = slice.as_ptr_range();
+some_ffi_function(r.start, r.end);
+```
+
+It also allows for things like
+
+```rust
+slice.as_ptr_range().contains(some_element)
```

Comment:
Is this well defined? In many other low level languages (okay, in C), it's only well defined if it returns true. I had heard this was the case for rust as well, but could have been misinformed.

Reply:
We could use the `wrapping_*` pointer operations to make this sound

Reply:
`contains` only uses comparisons though, which are safe, aren't they?

Reply:
> Is this well defined? In many other low level languages (okay, in C), it's only well defined if it returns true.

In C, pointer comparisons are defined only in cases where the pointers being compared originate from the same object. (Here, I think that object is commonly understood to mean "a stack or heap allocated value of some type") As Rust defines std::ptr::eq and `Ord` comparisons to be safe APIs (see here: https://doc.rust-lang.org/std/primitive.pointer.html), that means that they are guaranteed to be well defined.